### PR TITLE
 Host API usability improvements (Babylon)

### PR DESCRIPTION
**Pull request by Krxtopher**
*Issue #* SUMERIAN-11037

### Scope

The focus of this pull request is to encapsulate the ~350 lines of boilerplate code required to instantiate a host in BabylonJS (as demonstrated in our old example HTML file) to improve ease of use for users of the Hosts library. That boilerplate code is now encapsulated in a new module called **HostUtils** which exposes a simple `createHost()` function and a `getCharacterConfig()` function.

The old example app has been removed and two new examples have been added. These are the "hello-world-demo" and "gesture-demo" examples found under `/packages/amazon-sumerian-hosts-babylon/examples`.

#### Out of Scope

- Improving the implementation of the boilerplate code itself.
- Adding a similar HostUtils module to the Three.js implementation. That implementation will continue to use the verbose example file for now.
- Although I've added some minimal unit tests for HostUtils, most of the functionality must be manually tested by running the provided examples and observing expected host behavior.

### Files to review

You can ignore many of the auto-generated files. The files to review explicitly are:

- The following files under `/packages/amazon-sumerian-hosts-babylon/src/Babylon.js/`
  - HostUtils.js
  - index.js
- The new test spec: `/packages/amazon-sumerian-hosts-babylon/test/unit/HostUtils.spec.js`
- All files under `/packages/amazon-sumerian-hosts-babylon/examples`
- Any webpack config files
- Any package.json files
- Review the updated auto-generated API docs in a browser. It is not necessary to review the source code of these autogenerated files. See below for more detail on how to review the updated API docs.

### How to test

> **Prerequisites:**
> You must have a Cognito Identity Pool created in your AWS account. The unauthenticated IAM role for the pool (usually ending in the suffix "Unauth_Role") must have the following managed policies assigned to it:
>
> - AmazonPollyReadOnlyAccess
> - (optional) AmazonLexRunBotsOnly (optional for this PR)

Check out the repository.

In a terminal, `cd` to the repository's root directory.

Run `npm install`.

Update the following file to reflect your Cognito Identity Pool ID: `/packages/amazon-sumerian-hosts-babylon/examples/common/demo-credentials.js`

Run `npm run babylon-examples-serve` from the root of the repo. This will automatically start a web server and open a web browser to a listing of the available demos. Click either of the demos to test. If the hosts work properly, then the code is working!

### How to review API doc updates

Open the following file in a browser: `/docs/index.html`

Use the left-hand nav to navigate to "**Namespaces > babylonjs/HostUtils**".

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

## PR Update 1:
- I've removed generated "docs" files from the repo

### CODE REVIEW

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+
+  const animClips = {};
+  animLoadingResults.forEach((result) => {
+    animClips[result.clipGroupId] = result.clips;
+  });
+
+  // Load the gesture config file. This file contains options for splitting up
+  // each animation in gestures.glb into 3 sub-animations and initializing them
+  // as a QueueState animation.
+  const gestureConfig = await loadJson(gestureConfigUrl);
```
Comment (prestomation) [inline packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:153]: I don't love this pattern of loading json data. This seems like something that should be baked in at build-time to cut down on network calls/etc. Not sure it is worth addressing right now though

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import cognitoIdentityPoolId from './common/demo-credentials.js';
+
+let host;
+let scene;
+
+async function createScene() {
+  // Create an empty scene. IMPORTANT: Sumerian Hosts require use of the
+  // right-hand coordinate system!
+  scene = new BABYLON.Scene();
+  scene.useRightHandedSystem = true; // IMPORTANT for Sumerian Hosts!
```
Comment (prestomation) [inline packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:11]: You mentioned this before. Excuse my ignorance: what does this mean for customers with existing scenes? are they able to add a host to it?

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  // ===== Instantiate the Sumerian Host =====
+
+  // Edit the characterId if you would like to use one of
+  // the other pre-built host characters. Available character IDs are:
+  // "Cristine", "Fiona", "Grace", "Maya", "Jay", "Luke", "Preston", "Wes"
+  const characterId = 'Cristine';
+  const pollyConfig = { pollyVoice: 'Joanna', pollyEngine: 'neural' };
+  const characterConfig = HOST.HostUtils.getCharacterConfig('./assets/character-assets', characterId);
+  host = await HOST.HostUtils.createHost(scene, characterConfig, pollyConfig);
```
Comment (prestomation) [inline packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:31]: It's great that we moved so much critical code out of the sample app, but is something called HostUtils the right place? Should these go directly into HostObject? I'm not quite understanding the split between those. The HostObject looks like a more generic system, and HostUtils creates specific implementations that we know of as Preston/Wes/luke/etc? Should this be called SumerianHost or AmazonHost or something like that?

**Review by prestomation — COMMENTED** [review]
Do we intend to keep checking in generated files? I'd really prefer we did not do this.it bloats the repo size and makes commits hard to follow. The table of contents doesn't even work

File: packages/amazon-sumerian-hosts-babylon/examples/common/demo-utils.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+function setupSceneEnvironment(scene) {
+  // Create a simple environment.
+  const environmentHelper = scene.createDefaultEnvironment({
+    groundOpacity: 1,
+    groundShadowLevel: 0.3
+  });
+  environmentHelper.setMainColor(BABYLON.Color3.Teal());
+
+  // const light = new BABYLON.HemisphericLight("light", new BABYLON.Vector3(-0.3, 1, 0));
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/examples/common/demo-utils.js:9]: nit: do we want to remove the comment out line

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  const pollyConfig = { pollyVoice: 'Joanna', pollyEngine: 'neural' };
+  host = await HOST.HostUtils.createHost(scene, characterConfig, pollyConfig);
+
+  // Tell the host to always look at the camera.
+  const camera = scene.cameras[0];
+  host.PointOfInterestFeature.setTarget(camera);
+
+  // Enable shadows.
+  const light = scene.lights[0];
+  const shadowGenerator = new BABYLON.ShadowGenerator(2048, light);
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js:39]: nit: do we want to use shadowGenerator returned by DemoUtils.setupSceneEnvironment

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 2 above)
Comment (runpiw) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:11]: If we're doing this for gltf loading, it looks like the GLTFLoader is going to automatically hand this for us with flag GLTFLoaderCoordinateSystemMode default as AUTO

File: packages/amazon-sumerian-hosts-babylon/package.json [hunk 6]
```diff
@@ -17,7 +17,7 @@
     "test": "node ../../node_modules/karma/bin/karma start",
     "build": "webpack --config webpack.common.js",
     "build-test": "npm run build; npm run test",
-    "start": "webpack-dev-server --open --config webpack.dev.js",
+    "examples-serve": "webpack-dev-server --config webpack.dev.js",
     "build-start": "npm run build; npm run start",
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/package.json:21]: build-start also needs to change if we remove the start script

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+
+  return host;
+}
+
+/**
+ * @private
+ */
+async function initTextToSpeech() {
+  // Ensure services get initialized only once per session.
+  if (isTextToSpeechInitialized) return;
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:69]: Do we need this separate tracer variable or we could just use aws.TextToSpeechFeature.isReady

File: packages/amazon-sumerian-hosts-babylon/webpack.dev.js [hunk 8]
```diff
@@ -11,7 +11,7 @@ module.exports = merge(common, {
     modules: [path.resolve(__dirname, '../../node_modules'), 'node_modules'],
   },
   devServer: {
-    open: 'Google Chrome',
+    open: ['/examples'],
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/webpack.dev.js:14]: nit: just want to point out i think we do still have some of the integration test in babylon although i'm ok with ignoring them for now

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.html [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    <!-- Babylon.js dependencies -->
+    <script src="https://cdn.jsdelivr.net/npm/babylonjs@4.2.1/babylon.js"></script>
+    <script src="https://cdn.jsdelivr.net/npm/babylonjs-loaders@4.2.1/babylonjs.loaders.min.js"></script>
+
+    <!-- Sumerian Host dependencies -->
+    <script
+      type="text/javascript"
+      src="https://sdk.amazonaws.com/js/aws-sdk-2.645.0.min.js"
+    ></script>
+    <script type="text/javascript" src="../dist/host.babylon.js"></script>
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.html:28]: i think we could remove this src since we're using ES6 module for the script tag below which should also enable the live loading feature of webpack dev server

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 2 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:11]: @prestomation My understanding is that although _some_ existing scenes might be able to be adapted easily (static scenes in particular), many others won't. My hunch that a small change to our code could remove this right-hand coordinate system requirement, but I'm not sure. And I've considered it outside the scope of this PR.

Sounds like maybe Runping has some insight into a future potential fix?

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js (hunk 1 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:153]: Regarding inclusion of generated files, committing generated API docs definitely feels like a bad idea to bad idea to me as well and something I'd love to see changed. That said, changing it would require coming up with a different solution to how we host API docs using Github Pages. I don't know our options there very well. (On Ambit, our generated doc tool automatically creates a dedicated branch for the docs when we "publish" them and that branch is served by Github Pages.)

As far as committing the generated webpack bundle output, that also feels a bit akward. But I feel strongly we should give customers _some_ easy way to quickly grab the right bundle .js file without having to download and build the whole repo. What are our options there? Releases? Force them to download the NPM package and go diving into the "node_modules" folder? Something else?

Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:153]: Regarding the pattern I'm using for loading JSON data. Runtime loading of JSON is required because the correct JSON file(s) to load isn't necessarily known at build time. This function is used for loading things like "gesture" and "poi" (point-of-interest) config files which are different for each character (and for custom characters.) So if a customer wants to dynamically choose which character to instantiate at runtime, these files need to be loaded at runtime. Are you OK with me leaving this as-is?

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 3 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:31]: I waffled on this question myself and I'm happy to change it. And let me clarify, the HostUtils class can accommodate custom host characters as well as the "built-in" host characters. It does feel more like it should be a part of HostObject, but I was hesitant to bolt on 350 lines of code to that object. But it's probably a better choice. I'll make that change.

File: packages/amazon-sumerian-hosts-babylon/examples/common/demo-utils.js (hunk 4 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/common/demo-utils.js:9]: Thanks for catching this. I will fix.

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js (hunk 5 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js:39]: For now I'm choosing to use it in some demos and not others. My goal is that each demo should be slimmed down as much as possible to keep the user's focus on the specific feature we're trying to demonstrate. I don't consider shadows critical to showing off the gestures feature. I will leave this as-is for now.

File: packages/amazon-sumerian-hosts-babylon/package.json (hunk 6 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/package.json:21]: Good catch. I will fix.

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js (hunk 7 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:69]: I'll look into this. I had just kept the original implementation as it was and wasn't aware of the "isReady" property.

File: packages/amazon-sumerian-hosts-babylon/webpack.dev.js (hunk 8 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/webpack.dev.js:14]: Ah. I'll add a dedicated npm script for launching the integration tests.

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.html (hunk 9 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.html:28]: Although I happen to be using webpack-dev-server to make previewing the example site easier, I don't want a hard dependency on that server. I want to the examples to work if they're deployed to other servers as well (like Amplify hosting). Would your suggested change represent a hard dependency on webpack's server?

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 2 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:11]: BTW, the real-world impact of _not_ setting the coordinate system to right-handed is that the host's point-of-interest tracking works backwards. They look left when they should be looking right to track a target.

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js (hunk 1 above)
Comment (prestomation) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:153]: I don't have a lot of experience with github pages, but I know what you describe is a common pattern. It seems we should aim towards that. do we have a need for the docs to be checked in or updated right now until we get that addressed? My point being, if we know we want to do that(and I think we do) let's plan for it and stop checking in docs.

As for generated dist folder. When was the last time you github cloned a repo to get a generated file? Isn't `npm install babylon` the common route anyway? webdevs knows how to do that.


as for the json stuff. Yes, that is fine. I'm concerned about a potential pipelining problem on load, where the app needs to download the config, which downloads the gltf, which downloads the animations or whatever it is. It seems there is an opportunity to better bundle these things to optimize for load times. certainly not something we need to do right now, just worried about how it might be harder down the line.

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 2 above)
Comment (prestomation) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:11]: It seems to me we should swap the default in our hosts, so default babylon setup 'just works', but then make this configurable. otherwise we will have the opposite problem, customers using a right handed system(due to some other dependency they have) will then have a problem. we should default to whatever is more popular, which I assume is babylon's default.

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.html (hunk 9 above)
Comment (runpiw) [reply packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.html:28]: Sorry i didn't look clearly about how you structure files, it looks like we're just breaking down example source scripts. I'm ok to leave as this right now.

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js (hunk 1 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:153]: I like the suggestion of removing the generated docs for now. 

Regarding webdevs using npm install, you're probably right. Although in my projects I tend to link in libraries from from CDNs rather than compile those libraries into my code. I do this both for ease of development and also because it's more likely a new visitor to my site will already have a cached version of the CDN hosted lib, thus improving download times. But perhaps my practice is unusual given modern webdev practices. I'm happy to remove the compiled bundles from the repo, but will do that as a future PR since it may require some minor changes to our developer workflow (not sure.)

I agree that there's probably opportunity for optimization via bundling of JSON, 3D mesh, and animation files but will consider that out of scope for now.

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 2 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js:11]: I'm very much in agreement. I've created a new internal ticket for us to track this desired improvement. SUMERIAN-11086

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js (hunk 5 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js:39]: Ah. I now understand your comment better. In my next update to this PR you'll see I've updated the code to use the shadowGenerator returned by setupSceneEnvironment(). Great catch!

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js (hunk 1 above)
Comment (prestomation) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js:153]: Starting in the last coupld of years, the major browsers do this thing called 'Double-key'd HTTP caching', meaning that browser caching is key'd not only by the URL of the content, but also by the origin. This is for privacy reasons, but means that if you load two websites that both use the same version of babylon.js hosted on the same CDN(which itself will be a rare occurrence), the browser is still going to download and cache it twice: https://developers.google.com/web/updates/2020/10/http-cache-partitioning

Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js]: That's great insight!

File: packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js (hunk 3 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/hello-world-demo.js]: This issue is addressed in the update I just posted.

File: packages/amazon-sumerian-hosts-babylon/examples/common/demo-utils.js (hunk 4 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/common/demo-utils.js]: This issue is addressed in the update I just posted.

File: packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js (hunk 5 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/examples/gestures-demo.js]: This issue is addressed in the update I just posted.

File: packages/amazon-sumerian-hosts-babylon/package.json (hunk 6 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/package.json:21]: This issue is addressed in the update I just posted. Running `npm start` will now open a browser displaying the package root contents.

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js (hunk 7 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostUtils.js]: This issue is addressed in the update I just posted.

File: packages/amazon-sumerian-hosts-babylon/webpack.dev.js (hunk 8 above)
Comment (Krxtopher) [reply packages/amazon-sumerian-hosts-babylon/webpack.dev.js]: This issue is addressed in the update I just posted.

File: packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostObject.js [hunk 10]
```diff
@@ -84,6 +87,561 @@ class HostObject extends CoreHostObject {
 
     event.notifyObservers(value);
   }
+
+  /**
+   * @private
+   */
+  static async initTextToSpeech() {
```
Comment (runpiw) [inline packages/amazon-sumerian-hosts-babylon/src/Babylon.js/HostObject.js:94]: nit: This looks like a function which could sit in core HostObject class. Although, i guess in that case we should evaluate the necessity of letting user pass in their own aws service object or we always create for them like we do here.

**Comment by runpiw** [comment]
We should probably also update Readme to reflect the new script to run babylon example

**Comment by Krxtopher** [comment]
> We should probably also update Readme to reflect the new script to run babylon example

There are _lots_ of updates needed for the README. There's a whole epic devoted to updating documentation, so this will happen as part of that epic.

**Review by runpiw — APPROVED** [review]

File: .gitignore [hunk 11]
```diff
@@ -10,3 +10,4 @@ npm-debug.log
 .gitconfig
```
Comment (prestomation) [inline .gitignore:10]: Can we add the `demo-credentials.js` so devs have to deliberately add it to commit changes to it?

**Review by prestomation — APPROVED** [review]
